meta: add Profile management

Jan200101 opened the pull request:
meta PR to keep all the not yet PRable commits for profile management.

each and every one is going to be PR'd independenly and the profiles branch will get rebased after a merge.

GeckoEidechse commented:
Holy fuck this is so cool :DDDDD

Good thing I got distract with working on some other FlightCore feature today cause otherwise we would have done the same work at once lmao xD

Jan200101 commented on src-tauri/src/northstar/profile.rs:
```diff
@@ -0,0 +1,36 @@
+
+    let mut profiles: Vec<String> = Vec::new();
+
+    let pattern = format!("{}/*/mods", game_install.game_path);
+    dbg!(&pattern);
+    for e in glob::glob(&pattern).expect("Failed to read glob pattern") {
+        let path = e.unwrap();
+        let mut ancestors = path.ancestors();
+
+        ancestors.next();
+
+        let profile_path = std::path::Path::new(ancestors.next().unwrap());
```
Right now I only check for `/mods`, What are some other things to check for?

Right now I got:
- /mods
- /plugins
- /packages
- /Northstar.dll

Maybe the following?:
- logs
- runtime
- enabledmods.json
- placeholder.playerdata.pdata

GeckoEidechse replied:
Yeah, I'd say if any of those files are in there, treat it as a profile. Given that technically any folder could be a profile the condition for a profile more or less is "looks vaguely like it belongs to Norhtstar" and "isn't one of the game's existing folders like `r2` etc ^^

Maybe @BigSpice can tell us what he did for profile detection in VTOL :eyes:

Jan200101 commented on src-tauri/src/northstar/profile.rs:
```diff
@@ -0,0 +1,36 @@
+
+        let profile_path = std::path::Path::new(ancestors.next().unwrap());
+
+        profiles.push(profile_path.file_name().unwrap().to_os_string().into_string().unwrap());
```
yes, I know, map_err bla bla, I'll deal with it later.

GeckoEidechse commented on src-vue/src/plugins/store.ts:
```diff
@@ -334,6 +346,24 @@ export const store = createStore<FlightCoreStore>({
             );
+        },
+        async fetchProfiles(state: FlightCoreStore) {
+            let game_install = {
+                game_path: state.game_path,
+                profile: state.active_profile,
+                install_type: state.install_type
+            } as GameInstall;
```
One thing I wanted to do forever but always shyed from (cause it was kinda boring work) was to put the entire GameInstall object in the store as opposed to the individual struct members.

Not sure if that's something we wanna do before adding profile support of after given the it modifies the `GameInstall` struct. (obv separate PR either way). Thoughts?

Jan200101 replied:
Would simplify a lot of code and the game install storage.
I'd say before, I can slap together a PR that deals with it all.

Jan200101 commented:
> Good thing I got distract with working on some other FlightCore feature today cause otherwise we would have done the same work at once lmao xD

It would be fine, two people working on the same thing would give you two unique perspectives on the implementation, allowing you to cherry pick the best of both worlds :)

GeckoEidechse commented on src-tauri/src/northstar/profile.rs:
```diff
@@ -0,0 +1,36 @@
+    game_install: GameInstall,
+    profile: String
+) -> bool {
+    let profile_path = format!("{}/{}", game_install.game_path, profile);
+    let profile_dir = std::path::Path::new(profile_path.as_str());
+
+    return profile_dir.is_dir();
```
Maybe also make sure they aren't one of the game's existing folders to prevent installing into a game folder ^^

Jan200101 commented:
for the sake of not making this an absolute nightmare, I'll split this up into chunks and make new PRs for them.

Jan200101 commented:
This isn't the end of it, but all front facing functionality that would need to support Profiles is done.

Things to get done in the future:
- Migrate Launcher PRs and Mod PRs to the Profile system
- Fresh Profile creation

GeckoEidechse commented on src-vue/src/views/SettingsView.vue:
```diff
@@ -208,7 +343,6 @@ h3:first-of-type {
     margin-left: 50px;
 }
 
-
```
Randomly removed line :>

Jan200101 replied:
slipped in while trying to beat CSS into vue.
Will revert

GeckoEidechse commented on src-tauri/src/northstar/profile.rs:
```diff
@@ -0,0 +1,115 @@
+// These folders are part of Titanfall 2 and
+// should NEVER be used as a Profile
+const SKIP_PATHS: [&str; 7] = [
+    "__overlay",
```
I would add `___flightcore-temp` to this as well just to be save ^^

GeckoEidechse commented on src-tauri/src/northstar/profile.rs:
```diff
@@ -0,0 +1,115 @@
+    "save_data/",
+    "Northstar.dll",
+    "enabledmods.json",
+    "placeholder.playerdata.pdata",
```
If you wanna get ahead of the curve you can also add `LEGAL.txt` to the list ^^

https://github.com/R2Northstar/Northstar/issues/535

GeckoEidechse reviewed (commented):
So happy to see the progress :D

Left a few small comments.

With that said, I think we should try to still further reduce the diff a bit. In particular, as a next step I would splice out the Rust code for profile fetching/validation into a separate PR.
Add some buttons to test it in DevView and then we can merge that early as well.

After that it might also make sense to have the profile selection field in DevView for a bit so that we can make a release where contributors can basically test it in production.

Once it's considered stable we would just move the controls to settings for actual "release" ^^

Jan200101 commented:
squashed all commits
rebased onto master
separated every pragtical addition into separate commits to easily digest.

every single commit was tested to work (on Linux) individually and ran against clippy and the formatter.

This should make a review more feasible to do.

GeckoEidechse commented:
You're a saint. Will take a look tonight, thanks <3

Jan200101 commented:
little issue I have found:
switching profiles does not re-check the installation.

GeckoEidechse commented:
Didn't get to it, sorry :/

Given the nice commit structure, my plan is to just cherry-pick a the first few changes into separate PRs, so if you wanna get ahead of me feel free to just cherry-pick the first commit into a new PR and I will hopefully get to it tomorrow <3

Jan200101 commented:
#494

GeckoEidechse commented:
Superseded by #761
